# Ex 4.10 Part 1

# 01. Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02. Import data

#### 1. Import the data set you exported in Exercise 4.9 into a new Jupyter notebook. This should be the merged version of your data that contains the new customers column.

In [2]:
# Import final data from 4.9

path = r'C:\Users\Amy Yip\Desktop\Career Foundry Data Analytics\Achievement 4 Python\05-02-2022 Instacart Basket Analysis\02 Data\Prepared Data'

df_final = pd.read_pickle(os.path.join(path, 'df_ords_prods_all_v3.pkl'))

In [3]:
df_final.head()

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0  2539329       1                         1                   2   
1  2398795       1                         2                   3   
2   473747       1                         3                   3   
3  2254736       1                         4                   4   
4   431534       1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered  ... spending_flag  \
0                  1          0  ...   Low spender   
1                  1          1  ...   Low spender   
2                  1          1  ...   Low spender   
3                  1          1  ...   Low spender   
4                  1          1  ...   Low spender   

   median_days_since_prior_order   order_frequency_flag  gender    state age  \
0                           20.5  Non-frequent customer  Female  Alabama  31   
1                           20.5  Non-frequent customer  Female  Alabama  31   
2                           20.5  Non-frequent customer  Female  Alabama  31   
3                           20.5  Non-frequent customer  Female  Alabama  31   
4                           20.5  Non-frequent customer  Female  Alabama  31   

  date_joined  n_dependants fam_status  income  
0   2/17/2019             3    married   40423  
1   2/17/2019             3    married   40423  
2   2/17/2019             3    married   40423  
3   2/17/2019             3    married   40423  
4   2/17/2019             3    married   40423  

[5 rows x 30 columns]

In [4]:
pd.options.display.max_columns = None

In [5]:
df_final.shape

(32404859, 30)

In [6]:
df_final.dtypes

order_id                           object
user_id                            object
order_numbers_users_made             int8
orders_day_of_week                   int8
order_hour_of_day                    int8
days_since_prior_order            float64
prior_orders_made                    bool
product_id                          int64
add_to_cart_order                   int16
reordered                            int8
product_name                     category
aisle_id                            int64
department_id                       int64
prices                            float64
price_range_loc                  category
busiest_days                     category
busiest_period_of_day            category
max_order                            int8
loyalty_flag                     category
avg_spending                      float64
spending_flag                    category
median_days_since_prior_order     float64
order_frequency_flag             category
gender                           c

#### 2. Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

Personally identifiable information (PII) data within the customer dataframe was already addressed in Ex 4.9. Sensitive data first_name and surname columns were dropped from the dataframe to avoid being traced back to a particular person due to privacy regulation.

#### 3. The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.
* Use the region information in this Wikipedia article to create your column (you only need to create regions, not divisions).
* Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)

In [7]:
# Check states and value counts
df_final['state'].value_counts(dropna = False)

Pennsylvania            667082
California              659783
Rhode Island            656913
Georgia                 656389
New Mexico              654494
Arizona                 653964
North Carolina          651900
Oklahoma                651739
Alaska                  648495
Minnesota               647825
Massachusetts           646358
Wyoming                 644255
Virginia                641421
Missouri                640732
Texas                   640394
Colorado                639280
Maine                   638583
North Dakota            638491
Alabama                 638003
Kansas                  637538
Louisiana               637482
Delaware                637024
South Carolina          636754
Oregon                  636425
Arkansas                636144
Nevada                  636139
New York                635983
Montana                 635265
South Dakota            633772
Illinois                633024
Hawaii                  632901
Washington              632852
Mississi

#### Regions from Wikipedia https://en.wikipedia.org/wiki/List_of_regions_of_the_United_States

Region 1 (Northeast): Connecticut, Maine, Massachusetts, New Hampshire, Rhode Island, Vermont, New Jersey, New York, Pennsylvania

Region 2 (Midwest): Illinois, Indiana, Michigan, Ohio, Wisconsin, Iowa, Kansas, Minnesota, Missouri, Nebraska, North Dakota, South Dakota

Region 3 (South): Delaware, Florida, Georgia, Maryland, North Carolina, South Carolina, Virginia, Washington D.C., West Virginia, Alabama, Kentucky, Mississippi, Tennessee, Arkansas, Louisiana, Oklahoma, Texas

Region 4 (West): Arizona, Colorado, Idaho, Montana, Nevada, New Mexico, Utah, Wyoming, Alaska, California, Hawaii, Oregon, Washington

In [8]:
# Create new labels for regional segmentation based on states.

region = []

for state in df_final['state']:
    if (state == 'Connecticut') or (state == 'Maine') or (state == 'Massachusetts') or (state == 'New Hampshire') or (state == 'Rhode Island') or (state == 'Vermont') or (state == 'New Jersey') or (state == 'New York') or (state == 'Pennsylvania'):
        region.append('Northeast')
        
    elif (state == 'Illinois') or (state == 'Indiana') or (state == 'Michigan') or (state == 'Ohio') or (state == 'Wisconsin') or (state == 'Iowa') or (state == 'Kansas') or (state == 'Minnesota') or (state == 'Missouri') or (state == 'Nebraska') or (state == 'North Dakota')  or (state == 'South Dakota'):
        region.append('Midwest')
        
    elif (state == 'Arizona') or (state == 'Colorado') or (state == 'Idaho') or (state == 'Montana') or (state == 'Nevada') or (state == 'New Mexico')  or (state == 'Utah')  or (state == 'Wyoming') or (state == 'Alaska')  or (state == 'California') or (state == 'Hawaii') or (state ==  'Oregon') or (state == 'Washington'):
        region.append('West')
        
    else:
        region.append('South')

In [9]:
# Print region

region

['South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 'Midwest',
 '

In [10]:
# Create a new column "Region" containing these labels for regional segmentation

df_final['region'] = region

In [11]:
df_final['region'].value_counts()

South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
Name: region, dtype: int64

Check if regions total matches states total. The total should also match total rows of df_final dataframe:
10791885 + 8292913 + 7597325 + 5722736 = 32404859

In [12]:
len(df_final['region'])

32404859

In [13]:
len(df_final['state'])

32404859

Regions total does match states total. The total also matches the total rows of df_final dataframe.

In [14]:
# Determine whether there’s a difference in spending habits between the different U.S. regions. 
# (Hint: You can do this by crossing the variable you just created with the spending flag.)

# Create a crosstab between the 'region' column and  'spending_flag' column.

crosstab = pd.crosstab(df_final['region'], df_final['spending_flag'], dropna = False)

In [15]:
# Check results of crosstab

crosstab

spending_flag  High spender  Low spender
region                                  
Midwest              155975      7441350
Northeast            108225      5614511
South                209691     10582194
West                 160354      8132559

In [16]:
# % of high spenders in the population of Midwest region

Midwest_high_spenders_percent = (155975/(155975+7441350))*100

In [17]:
Midwest_high_spenders_percent

2.053025242437305

In [18]:
# % of high spenders in the population of Northeast region

Northeast_high_spenders_percent = (108225/(108225+5614511))*100

In [19]:
Northeast_high_spenders_percent

1.8911408808653762

In [20]:
# % of high spenders in the population of South region

South_high_spenders_percent = (209691/(209691+10582194))*100

In [21]:
South_high_spenders_percent

1.9430433144904713

In [22]:
# % of high spenders in the population of West region

West_high_spenders_percent = (160354/(160354+8132559))*100

In [23]:
West_high_spenders_percent

1.9336269414619447

The spending habits between the four regions do not vary much. In fact, they are very similar, with highest spending population percentage of 2.05% at the Midwest region and lowest spending population percentage of 1.89% at the Northeast region.

In [24]:
df_final.shape

(32404859, 31)

In [25]:
df_final.head()

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0  2539329       1                         1                   2   
1  2398795       1                         2                   3   
2   473747       1                         3                   3   
3  2254736       1                         4                   4   
4   431534       1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         10   
1  Mid-range product    Slowest days           Most orders         10   
2  Mid-range product    Slowest days           Most orders         10   
3  Mid-range product    Slowest days           Most orders         10   
4  Mid-range product    Slowest days           Most orders         10   

   loyalty_flag  avg_spending spending_flag  median_days_since_prior_order  \
0  New customer      6.367797   Low spender                           20.5   
1  New customer      6.367797   Low spender                           20.5   
2  New customer      6.367797   Low spender                           20.5   
3  New customer      6.367797   Low spender                           20.5   
4  New customer      6.367797   Low spender                           20.5   

    order_frequency_flag  gender    state  age date_joined  n_dependants  \
0  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
1  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
2  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
3  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
4  Non-frequent customer  Female  Alabama   31   2/17/2019             3   

  fam_status  income region  
0    married   40423  South  
1    married   40423  South  
2    married   40423  South  
3    married   40423  South  
4    married   40423  South

#### 4. The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

In [26]:
# Setting new flag - Low-activity customer

df_final.loc[df_final['max_order'] < 5, 'activity_flag']= 'Low-activity customer'
df_final.loc[df_final['max_order'] >= 5, 'activity_flag']= 'Normal-activity customer'

In [27]:
# Check activity_flag

df_final['activity_flag'].value_counts(dropna = False)

Normal-activity customer    30964564
Low-activity customer        1440295
Name: activity_flag, dtype: int64

In [28]:
# Check total 
30964564+1440295

32404859

In [29]:
# Create a subset of dataframe excluding the low-activity customers

df_normal_activity = df_final[df_final['activity_flag'] == 'Normal-activity customer']

In [30]:
df_normal_activity

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0         2539329       1                         1                   2   
1         2398795       1                         2                   3   
2          473747       1                         3                   3   
3         2254736       1                         4                   4   
4          431534       1                         5                   4   
...           ...     ...                       ...                 ...   
32404850  3308056  106143                        22                   4   
32404851  2988973  106143                        23                   2   
32404852      930  106143                        24                   6   
32404853   467253  106143                        25                   6   
32404854   156685  106143                        26                   4   

          order_hour_of_day  days_since_prior_order  prior_orders_made  \
0                         8                     NaN               True   
1                         7                    15.0              False   
2                        12                    21.0              False   
3                         7                    29.0              False   
4                        15                    28.0              False   
...                     ...                     ...                ...   
32404850                 20                    10.0              False   
32404851                 22                     5.0              False   
32404852                 12                     4.0              False   
32404853                 16                     7.0              False   
32404854                 23                     5.0              False   

          product_id  add_to_cart_order  reordered  \
0                196                  1          0   
1                196                  1          1   
2                196                  1          1   
3                196                  1          1   
4                196                  1          1   
...              ...                ...        ...   
32404850       19675                  1          1   
32404851       19675                  1          1   
32404852       19675                  1          1   
32404853       19675                  1          1   
32404854       19675                  1          1   

                         product_name  aisle_id  department_id  prices  \
0                                Soda        77              7     9.0   
1                                Soda        77              7     9.0   
2                                Soda        77              7     9.0   
3                                Soda        77              7     9.0   
4                                Soda        77              7     9.0   
...                               ...       ...            ...     ...   
32404850  Organic Raspberry Black Tea        94              7    10.7   
32404851  Organic Raspberry Black Tea        94              7    10.7   
32404852  Organic Raspberry Black Tea        94              7    10.7   
32404853  Organic Raspberry Black Tea        94              7    10.7   
32404854  Organic Raspberry Black Tea        94              7    10.7   

            price_range_loc    busiest_days busiest_period_of_day  max_order  \
0         Mid-range product  Regularly busy           Most orders         10   
1         Mid-range product    Slowest days           Most orders         10   
2         Mid-range product    Slowest days           Most orders         10   
3         Mid-range product    Slowest days           Most orders         10   
4         Mid-range product    Slowest days           Most orders         10   
...                     ...             ...                   ...        ...   
32404850  Mid-range product    Slowest days           Most orders         26   
32404851  Mid-range product  Regularly bu

In [31]:
# Check value counts of 'activity_flag' column from the subset.

df_normal_activity['activity_flag'].value_counts()

Normal-activity customer    30964564
Name: activity_flag, dtype: int64

The subset df_normal_activity dataframe value count matches the 'Normal-activity customer' column orginal df_final dataframe.

In [32]:
df_normal_activity.shape

(30964564, 32)

In [33]:
df_final.shape

(32404859, 32)

In [34]:
df_normal_activity.head()

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0  2539329       1                         1                   2   
1  2398795       1                         2                   3   
2   473747       1                         3                   3   
3  2254736       1                         4                   4   
4   431534       1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         10   
1  Mid-range product    Slowest days           Most orders         10   
2  Mid-range product    Slowest days           Most orders         10   
3  Mid-range product    Slowest days           Most orders         10   
4  Mid-range product    Slowest days           Most orders         10   

   loyalty_flag  avg_spending spending_flag  median_days_since_prior_order  \
0  New customer      6.367797   Low spender                           20.5   
1  New customer      6.367797   Low spender                           20.5   
2  New customer      6.367797   Low spender                           20.5   
3  New customer      6.367797   Low spender                           20.5   
4  New customer      6.367797   Low spender                           20.5   

    order_frequency_flag  gender    state  age date_joined  n_dependants  \
0  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
1  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
2  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
3  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
4  Non-frequent customer  Female  Alabama   31   2/17/2019             3   

  fam_status  income region             activity_flag  
0    married   40423  South  Normal-activity customer  
1    married   40423  South  Normal-activity customer  
2    married   40423  South  Normal-activity customer  
3    married   40423  South  Normal-activity customer  
4    married   40423  South  Normal-activity customer

In [35]:
df_final.head()

order_id user_id  order_numbers_users_made  orders_day_of_week  \
0  2539329       1                         1                   2   
1  2398795       1                         2                   3   
2   473747       1                         3                   3   
3  2254736       1                         4                   4   
4   431534       1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         10   
1  Mid-range product    Slowest days           Most orders         10   
2  Mid-range product    Slowest days           Most orders         10   
3  Mid-range product    Slowest days           Most orders         10   
4  Mid-range product    Slowest days           Most orders         10   

   loyalty_flag  avg_spending spending_flag  median_days_since_prior_order  \
0  New customer      6.367797   Low spender                           20.5   
1  New customer      6.367797   Low spender                           20.5   
2  New customer      6.367797   Low spender                           20.5   
3  New customer      6.367797   Low spender                           20.5   
4  New customer      6.367797   Low spender                           20.5   

    order_frequency_flag  gender    state  age date_joined  n_dependants  \
0  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
1  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
2  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
3  Non-frequent customer  Female  Alabama   31   2/17/2019             3   
4  Non-frequent customer  Female  Alabama   31   2/17/2019             3   

  fam_status  income region             activity_flag  
0    married   40423  South  Normal-activity customer  
1    married   40423  South  Normal-activity customer  
2    married   40423  South  Normal-activity customer  
3    married   40423  South  Normal-activity customer  
4    married   40423  South  Normal-activity customer

In [32]:
# Export subset df_normal_activity

df_normal_activity.to_pickle(os.path.join(path, 'df_ords_prods_all_normal_activity_cust.pkl'))

In [33]:
# Export updated df_final

df_final.to_pickle(os.path.join(path,'df_ords_prods_all_Ex4.10.1.pkl'))

#### 5. The marketing and business strategy units at Instacart want to create more-relevant marketing strategies for different products and are, thus, curious about customer profiling in their database. Create a profiling variable based on age, income, certain goods in the “department_id” column, and number of dependents. You might also use the “orders_day_of_the_week” and “order_hour_of_day” columns if you can think of a way they would impact customer profiles. (Hint: As an example, try thinking of what characteristics would lead you to the profile “Single adult” or “Young parent.”)

See part 2 (using all df).

#### 6. Create an appropriate visualization to show the distribution of profiles.

See part 3 and 4.

#### 7. Aggregate the max, mean, and min variables on a customer-profile level for usage frequency and expenditure.

See part 5.

#### 8. Compare your customer profiles with regions and departments. Can you identify a link between them? Customers may demonstrate different behaviors across regions, which could make this information important.

See response in Excel spreadsheet.

#### 9. Produce suitable charts to illustrate your results and export them in your project folder.

#### 10. Tidy up your script, export your final data set, and save your notebook.